src/main/fc/core.c

```diff
@@ -1091,6 +1099,18 @@ FAST_CODE void taskMainPidLoop(timeUs_t currentTimeUs)
     if (lockMainPID() != 0) return;
 #endif
 
+#ifdef USE_DSHOT_TELEMETRY
+    static bool bidirDshotInited = false;
+    if (currentTimeUs > 5000000 && !bidirDshotInited) {
+        bidirDshotInited = true;
+        pwmWriteDshotCommand(
```

Comment:
You need to send a Dshot disarm command before you send any other command.

Reply:
@mikeller, I don't think a disarm dshot command exists or is sent by us in any other situation. Happy to add one if you can point me to it. But if it does exist wouldn't we send it on FC startup in case the motors are still running?

A delay after startup is needed - I found that sending the command while the esc's still beep did not work. Happy to move it to a place you find more appropriate.

Reply:
https://github.com/betaflight/betaflight/blob/131e32c142cf4da3ffddad0078d32579eda7c5ce/src/main/drivers/pwm_output.h#L44

Sorry, my bad, it's called `MOTOR_STOP`, but anyhow, Dshot commands will only be accepted if this (and not a motor speed value) has been sent as the last speed setting.

In any case, even waiting 5 seconds will not necessarily guarantee that the ESCs are ready (think when the flight controller is powered from USB while bench testing, and a battery is plugged in at a later stage).

Your best bet is probably to add the commands to `tryArm()` (https://github.com/betaflight/betaflight/blob/131e32c142cf4da3ffddad0078d32579eda7c5ce/src/main/fc/core.c#L367), just like what is done for reversing the motor directions for crash flip. This will also ensure that the motors are stopped / disarmed at the time.

Reply:
Right, I did see that. But we never send that command and sskaugs spec says it's not implemented. The spec says that commands can be send when the motors are stopped, but I think that's just when they don't turn vs having sent the MOTOR_STOP command.

I already do send the command in tryArm as well. The delayed send after startup allows testing the feature without arming which has proved valuable in testing.

Reply:
But agreed, when the battery is plugged in late that doesn't work. One way to address it is to periodically send it when unarmed. That might actually be a good idea. Or send it when the battery is attached, but we might not always have a reliable indicator for that.

Reply:
> But we never send that command and sskaugs spec says it's not implemented. 

That is just plain wrong:

https://github.com/betaflight/betaflight/blob/131e32c142cf4da3ffddad0078d32579eda7c5ce/src/main/flight/mixer.c#L378

Also, just because some closed source implementation of Dshot is broken, we should not strive to repeat the brokenness.

Just sending the command when arming should be good enough - the code sending it after the initialisation could be made conditional, allowing debugging with debug builds - but even then it should be done with a dispatch call (`src/main/fc/dispatch.c`) from `init()`, and not in the PID loop.

Reply:
Thank you, thank you, thank you! I could just not find it since it doesn't use the dshot cmd definitions in pwm_output.h but introduces a separate new define 😄.

It does look like that command gets sent continuously when the quad is disarmed though by the mixer so I'm not sure it makes any sense to send it again explicitly when issuing the telemetry cmd during arming or after startup? Please clarify.

I hadn't yet encountered the dispatch facility. Thanks for pointing it out!

Reply:
Inconsistency addressed in #7281.

You are correct that it does not make sense to explicitly issue the disarm command when sending commands in `tryArm()` since the craft is by definition disarmed at that point. This is why `tryArm()` is a good place for issuing commands.

However, in the case of the command that is issued after a delay of some seconds after initialisation, there is no guarantee that the craft is still disarmed at that point - assuming a board that initialises fast, and a quick user, the craft could be armed by then. (One way around this will be to check if the craft is armed, and if so, since the command was issued during arming, just not doing anything.)

Reply:
Right - though I've found it to be just about impossible to arm within 5s of boot and I already send the command on tryArm, so if one does manage to arm that quickly the mode is already set and the command will just be ignored.

I've moved this call to a dispatch set in init and made it contingent on the craft no being armed as suggested. Also added a comment to potentially remove once the feature is stable.